File: 324 Wiggle Sort II/324 Wiggle Sort II.py
"""
Given an integer array nums, reorder it such that nums[0] < nums[1] > nums[2] < nums[3]....
You may assume the input array always has a valid answer.

Example 1:

Input: nums = [1,5,1,1,6,4]
Output: [1,6,1,5,1,4]
Explanation: [1,4,1,5,1,6] is also accepted.

Example 2:

Input: nums = [1,3,2,2,3,1]
Output: [2,3,1,3,1,2]
"""
class Solution:
    def wiggleSort(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        nums.sort()
        half = len(nums[::2])
        nums[::2], nums[1::2] = nums[:half][::-1], nums[half:][::-1]

File: 324 Wiggle Sort II/324 Wiggle Sort II.go
func wiggleSort(nums []int)  {    
    wiggle(nums)
}

func wiggle(nums []int) {
    n := len(nums)

    sort.Slice(nums, func(i, j int) bool {
        return nums[i] < nums[j]
    })

    i, j, k := 0, 0, n - 1
    median := nums[n/2]

    for j <= k {
        nj, aj := A(nums, j)
        if nj > median {
            _, ai := A(nums, i)
            nums[ai], nums[aj] = nums[aj], nums[ai]
            i++
            j++
        } else if nj < median {
            _, ak := A(nums, k)
            nums[aj], nums[ak] = nums[ak], nums[aj]
            k--
        } else {
            j++
        }
    }
}

func A(nums[]int, i int) (int, int) {
    idx := (1 + 2*i) % (len(nums) | 1)
    return nums[idx], idx
}
